File: algorithms_toolbox/week3_greedy_algorithms/2_maximum_value_of_the_loot/FractionalKnapsack.java

import java.util.Scanner;

public class FractionalKnapsack {
    private static double getOptimalValue(int capacity, int[] values, int[] weights) {
        double value = 0;
        int n = values.length;
        //write your code here
        boolean bool[] = new boolean[n];
        for (int i = 0; i < n; i++) {
            bool[i] = true;
        }
        int index = 0;
        while (capacity > 0) {
            double max = 0;
            for (int i = 0; i < n; i++) {
                if (!bool[i])
                    continue;
                double ratio;
                ratio = (float) values[i] / (float) weights[i];
                if (ratio > max) {
                    max = ratio;
                    index = i;
                }
            }
            int weight = weights[index];
            if (capacity >= weight && bool[index]) {
                capacity -= weight;
                value += values[index];
            } else if (bool[index]){
                value += (float) (values[index] * capacity) / (float) weight;
                capacity = 0;
            }
            bool[index] = false;
            System.out.println(capacity);
        }

        return value;
    }

    public static void main(String args[]) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int capacity = scanner.nextInt();
        int[] values = new int[n];
        int[] weights = new int[n];
        for (int i = 0; i < n; i++) {
            values[i] = scanner.nextInt();
            weights[i] = scanner.nextInt();
        }
        System.out.printf("%.4f",getOptimalValue(capacity, values, weights));
    }
}

File: algorithms_toolbox/week3_greedy_algorithms/2_maximum_value_of_the_loot/fractional_knapsack.py
# Uses python3
import sys

def get_optimal_value(capacity, weights, values):
    value = 0.

    # write your code here
    ratios = list()
    for i in range(len(values)):
        ratios.append(values[i]/weights[i])
    while capacity > 0:
        if len(ratios) == 0:
            break
        max_ratio = max(ratios)
        max_index = ratios.index(max_ratio)
        max_val = values[max_index]
        max_weight = weights[max_index]

        if max_weight > capacity:
            value += max_val*(capacity/max_weight)
            break

        ratios.remove(max_ratio)
        value += max_val
        capacity -= max_weight
        #print(ratios, "ratio", max_ratio, "max_ratio", max_ratio, "max_weight", max_weight)
    return value


if __name__ == "__main__":
    data = list(map(int, sys.stdin.read().split()))
    n, capacity = data[0:2]
    values = data[2:(2 * n + 2):2]
    weights = data[3:(2 * n + 2):2]
    opt_value = get_optimal_value(capacity, weights, values)
    print("{:.4f}".format(opt_value))
